docs/design/2020-06-27-non-recursive-common-table-expression.md

```diff
@@ -0,0 +1,85 @@
+
+# Proposal: Non-Recursive Common Table Expression
+
+- Author(s):     [Pingyu](https://github.com/pingyu) (Ping Yu)
+- Last updated:  2020-06-29
+- Discussion at: https://github.com/pingcap/tidb/issues/17472
+
+## Abstract
+
+This proposal proposes to support the __Non-recursive Common Table Expression (CTE)__.
+
+## Background
+
+From MySQL User Document, [13.2.15 WITH (Common Table Expressions)](https://dev.mysql.com/doc/refman/8.0/en/with.html):
+> A common table expression (CTE) is a named temporary result set that exists within the scope of a single statement and that can be referred to later within that statement, possibly multiple times. The following discussion describes how to write statements that use CTEs.
+
+E.g.
+```sql
+WITH
+  cte1 AS (SELECT a, b FROM table1),
+  cte2 AS (SELECT c, d FROM table2)
+SELECT b, d FROM cte1 JOIN cte2
+WHERE cte1.a = cte2.c;
+```
+
+## Proposal
+
+We design two strategies for CTE: __Inline__ & __Materialization__:
```

Comment:
Looks like both approaches have shortcomings: `Inline` would execute the CTE multiple times while `Materialization` would disable potential optimizations. A more appropriate approach would be that, we introduce a new operator like `CTEReader` which follows `Materialization` strategy basically, and in planner we decide whether to expand a specific `CTEReader`, i.e, switch to `Inline` strategy, according to cost estimation. It is pretty easy to incorporate this approach into Cascades planner, but difficult for current default planner.

Reply:
I agree.
And ["Optimization of Common Table Expressions in MPP Database Systems"](http://www.vldb.org/pvldb/vol8/p1704-elhelw.pdf) introduces a very similar design, including `CTEReader` (called `CTEConsumer` in the paper) and use of cascades planner.

So how is the cascades planner going on ? Will it be the default one in next major version ?

Reply:
> So how is the cascades planner going on ? Will it be the default one in next major version ?

I don't think so, we can focus it on the current planner.

Reply:
Does MySQL support hint for enforcing a specific strategies.

Reply:
Yes. See [8.2.2.4 Optimizing Derived Tables, View References, and Common Table Expressions with Merging or Materialization](https://dev.mysql.com/doc/refman/8.0/en/derived-table-optimization.html).

> The following means are available to influence whether the optimizer attempts to merge derived tables, view references, and common table expressions into the outer query block:

> * The MERGE and NO_MERGE optimizer hints can be used. They apply assuming that no other rule prevents merging. See Section 8.9.3, “Optimizer Hints”.

> * Similarly, you can use the derived_merge flag of the optimizer_switch system variable. See Section 8.9.2, “Switchable Optimizations”.

E.g.
```sql
mysql root@localhost:test1> explain with cte(k, v) as  
                            ( select i, j from test1.t1 )  
                            select * from cte; 
                                                                                                                                                                                                              
+----+-------------+-------+------------+------+---------------+--------+---------+--------+------+----------+--------+
| id | select_type | table | partitions | type | possible_keys | key    | key_len | ref    | rows | filtered | Extra  |
+----+-------------+-------+------------+------+---------------+--------+---------+--------+------+----------+--------+
| 1  | SIMPLE      | t1    | <null>     | ALL  | <null>        | <null> | <null>  | <null> | 42   | 100.0    | <null> |
+----+-------------+-------+------------+------+---------------+--------+---------+--------+------+----------+--------+
1 row in set
Time: 0.014s
mysql root@localhost:test1>                                                                                                                                                                                   
mysql root@localhost:test1> explain with cte(k, v) as  
                            ( select i, j from test1.t1 )  
                            select /*+ NO_MERGE(cte) */ * from cte; 
                                                                                                                                                                                                              
+----+-------------+------------+------------+------+---------------+--------+---------+--------+------+----------+--------+
| id | select_type | table      | partitions | type | possible_keys | key    | key_len | ref    | rows | filtered | Extra  |
+----+-------------+------------+------------+------+---------------+--------+---------+--------+------+----------+--------+
| 1  | PRIMARY     | <derived2> | <null>     | ALL  | <null>        | <null> | <null>  | <null> | 42   | 100.0    | <null> |
| 2  | DERIVED     | t1         | <null>     | ALL  | <null>        | <null> | <null>  | <null> | 42   | 100.0    | <null> |
+----+-------------+------------+------------+------+---------------+--------+---------+--------+------+----------+--------+
2 rows in set
Time: 0.015s
mysql root@localhost:test1>  

```

Reply:
Suggest adding hint policy into this proposal.

Reply:
Hint policy added. PTAL~
@SunRunAway